Next100 shielding revision

gondiaz opened the pull request:
This PR revisits and updates the shielding geometry and generators. The dimensions and components were gently provided by Jordi Torrent and are shown in

[Shielding.pdf](https://github.com/next-exp/nexus/files/6554546/101-Castle.and.new.dimensions.pdf)
[pedastal.pdf](https://github.com/next-exp/nexus/files/6602464/pedastal.pdf)

Besides the relevant change of the new dimensions, two extra changes are made in this revision:
- the steel shielding in the laterals is enlarged, as shown in the drawings. The steel roof is also positioned above the inner steel box.
- solves a bug in the distribution of the vertex generator in the steel shielding structure composed by the roof and the top beams.
I have also tryed to rewrite the code to be more easily readable.

gondiaz committed:
Rewrite shielding geometry with correct dimensions

gondiaz committed:
Rewrite generators

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -36,27 +36,25 @@ namespace nexus {
 
   {
-
     /// Shielding is compound by two boxes, the external made of lead,
```
I would change the description to something like "The shielding is made of two boxes. The external one is made of lead and the internal one is made of a mix of steel and titanium. The steel beam structure is placed inside the lead." Is it correct that the internal box is made of steel and titanium?

I would also add a brief explanation about how we simulate the beam structure.

gondiaz replied:
Yes, following Jordi's drawing the internal steel box is made of "Stainless steel AISI-316Ti". Shoudl we inplement the box with this material? Not sure if it exists in G4. On the other hand, the castle structure is made of steel S-275.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -273,13 +262,15 @@ namespace nexus {
+    G4double lat_beam_x_vol = (shield_x_ + 2.*lead_thickness_ + 2.*steel_thickness_)*(lead_thickness_ - beam_base_thickness_)*beam_base_thickness_;
+    perc_struc_x_vol_    = 4*lat_beam_x_vol/struct_top_vol;
 
     // std::cout<<"SHIELDING LEAD VOLUME:\t"<<lead_box_solid->GetCubicVolume()<<std::endl;
```
I would start using a verbosity configuration parameters in the geometry classes (following what @jmunozv already did with NextFlex), which, if set to `true` prints out these comments. I would add units and a bit more verbosity to explain what is printed.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -70,8 +68,6 @@ namespace nexus {
-    //   G4Box* shielding_box_solid = new G4Box("SHIELD_BOX", shield_x_/2., shield_y_/2., shield_z_/2.);
 
     // LEAD BOX   ///////////
     lead_x_ = shield_x_ + 2. * steel_thickness_ + 2. * lead_thickness_;
```
The `y` dimension has to take into account the thickness of the beam base.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -36,27 +36,25 @@ namespace nexus {
-    roof_z_separation_ (755. *mm), //distance between 1st and 2nd roof beams
-    front_x_separation_ (154.*mm), //distance between the two front beams
+    beam_base_thickness_ (4.   * mm),
+    lateral_z_separation_(1010.* mm), //distance between the two lateral beams
```
Are these the distances between the centres of the beams or between the sides? Please, specify. If it's the second case, the union of solids below needs to be reviewed.

gondiaz replied:
From Jordi's drawings I think it is between the beam centers. If it were not the case, the total distance (see lateral view in slide 5) would not sum up, since we had to add two times the beam thickness.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -86,127 +82,120 @@ namespace nexus {
+    // auxiliar positions used in translations
+    G4double lat_beam_x   = shield_x_/2. + steel_thickness_ + lead_thickness_/2.;
+    G4double front_beam_z = shield_z_/2. + steel_thickness_ + lead_thickness_/2.;
+    G4double top_beam_y   = shield_y_/2. + steel_thickness_ + (lead_thickness_ + beam_base_thickness_)/2.;
```
I would say that `beam_base_thickness_` has to be added, not half.

paolafer replied:
The same should be reviewed in all the dimensions and placements regarding the `y` coordinate.

gondiaz committed:
Clarify shielding description

gondiaz committed:
Add verbosity parameter

gondiaz committed:
Add extra beam thickness in box y direction

jmalbos commented on source/geometries/Next100Shielding.cc:
```diff
@@ -36,46 +36,45 @@ namespace nexus {
-    shield_y_ (225.6 * cm),
-    shield_z_ (258.5 * cm), // 253.0 * cm before May 12, 2017
+    shield_x_ (158.  * cm),
+    shield_y_ (166.  * cm),
```
According to the drawings, this should be 166.6 cm, right?

gondiaz replied:
Right

jmalbos commented on source/geometries/Next100Shielding.cc:
```diff
@@ -36,46 +36,45 @@ namespace nexus {
-    shield_x_ (155.  * cm),
-    shield_y_ (225.6 * cm),
-    shield_z_ (258.5 * cm), // 253.0 * cm before May 12, 2017
+    shield_x_ (158.  * cm),
```
I thought we had decided to start using braces for initialisation: `shield_x_ {158. * cm}`

gondiaz committed:
Fix shield_y_ size

gondiaz committed:
Cosmetics, use bracket for initialization

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -86,127 +85,122 @@ namespace nexus {
+    G4double front_beam_z = shield_z_/2. + steel_thickness_ + lead_thickness_/2.;
+    G4double top_beam_y   = (shield_y_-beam_thickness_)/2. + steel_thickness_ + beam_thickness_ + lead_thickness_/2.;
+    G4double lat_beam_y   = -(lead_thickness_ + beam_thickness_)/2.;
+    G4double roof_y       = (shield_y_-beam_thickness_)/2. + steel_thickness_ + beam_thickness_/2.;
```
Isn't it `roof_y       = shield_y_/2. + steel_thickness_ + beam_thickness_/2.`?

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -86,127 +85,122 @@ namespace nexus {
+    // auxiliar positions used in translations
+    G4double lat_beam_x   = shield_x_/2. + steel_thickness_ + lead_thickness_/2.;
+    G4double front_beam_z = shield_z_/2. + steel_thickness_ + lead_thickness_/2.;
+    G4double top_beam_y   = (shield_y_-beam_thickness_)/2. + steel_thickness_ + beam_thickness_ + lead_thickness_/2.;
```
Isn't it `top_beam_y   = shield_y_/2. + steel_thickness_ + beam_thickness_ + lead_thickness_/2.` ?

gondiaz replied:
No, because now the lead box center is "displaced" by `beam_thickness_/2` because the `lead_y_` dimension has and extra term of ` beam_thickness_`

paolafer replied:
You are right, it's correct.

gondiaz committed:
Remove lower side of steel box

Also, set beam_thickness_ and steel_thickness_ to a large value for
visualization purpuses in geometry development. Its correct values
will be added once the geometry is finished.

gondiaz committed:
Add pedestal beams

gondiaz committed:
Add beams correct dimensions

gondiaz committed:
Add two beam thickness in steel structure

gondiaz committed:
Add pedestal material

gondiaz committed:
Add pedestal generator

gondiaz committed:
Add correct pedestal x dimension

gondiaz committed:
Rename lateral to front and correct pedestal_x

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -86,128 +93,182 @@ namespace nexus {
+                      pedestal_beam_lat_logic, "PEDESTAL_BEAM_LAT_2", lead_box_logic, false, 0);
 
     // SETTING VISIBILITIES   //////////
     if (visibility_) {
```
I would add the visibility behaviour of the new pedestal volumes. We can set the same one as the rest of the structure, for instance, and make them invisible if visibility is zero.

gondiaz committed:
Add pedestal roof (semi-T) and lateral beams

gondiaz committed:
Solve wrong naming and set pedestal to invisible

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -217,74 +311,135 @@ namespace nexus {
+      roof_logic    ->SetVisAttributes(antiox_col);
       lat_beam_logic->SetVisAttributes(antiox_col);
-      struct_logic->SetVisAttributes(antiox_col);
+      struct_logic  ->SetVisAttributes(antiox_col);
```
`front_beam_logic` should be also added here: in the current version of the code, the front beams are visible even when `visibility == 0`.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -217,74 +311,135 @@ namespace nexus {
-      roof_logic->SetVisAttributes(G4VisAttributes::Invisible);
-      struct_logic->SetVisAttributes(G4VisAttributes::Invisible);
-      lat_beam_logic->SetVisAttributes(G4VisAttributes::Invisible);
+      roof_logic     ->SetVisAttributes(G4VisAttributes::Invisible);
```
Same here: add the same color for the front beams.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -217,74 +311,135 @@ namespace nexus {
+
+      G4double inner_vol = air_box_solid  ->GetCubicVolume();
+      G4double steel_vol = steel_box_solid->GetCubicVolume() - inner_vol;
+      G4double lead_vol  = lead_box_solid ->GetCubicVolume() - steel_vol;
```
Fix the definition of `lead_vol`.

gondiaz committed:
Fix front beam visibility

gondiaz committed:
Fix lead volume computation

gondiaz committed:
Rename NextNew pedestal vertex generator

Fix name in NextNewPedestal class

gondiaz committed:
Fix shielding steel vertex generator

gondiaz committed:
Add bubble seal vertex generator

gondiaz committed:
Add EDPM seal generator

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -33,49 +33,65 @@ namespace nexus {
+    pedestal_top_x_    {1384. * mm},
+    support_beam_dist_ {1027. * mm},
+    support_front_dist_{732.  * mm},
+    pedestal_lateral_beam_thickness {8. * mm},
```
Add trailing underscore in this line and the following, according to the nexus conventions.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -217,74 +315,165 @@ namespace nexus {
+                                           G4ThreeVector(pedestal_x_pos, pedestal_y_pos, 0.), 0);
+
+    // pedestal roof
+    G4double ped_roof_gen_x_ = pedestal_top_x_/2. + pedestal_roof_thickness/2.;
```
Remove trailing underscore from the variable defined in lines 404, 405, 406, according to nexus conventions.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -217,74 +315,165 @@ namespace nexus {
+    bubble_seal_lateral_gen_ = new BoxPointSampler(bubble_seal_thickness, bubble_seal_thickness, pedestal_lateral_length, 0.,
+                                                   G4ThreeVector(0., ped_gen_y_, 0.), 0);
+
+    G4double bubble_front_vol_   = 2*(pedestal_x_ + 2*pedestal_lateral_beam_thickness)*(bubble_seal_thickness)*(bubble_seal_thickness);
```
Remove trailing underscores, as above.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -217,74 +315,165 @@ namespace nexus {
+    edpm_seal_lateral_gen_ = new BoxPointSampler(edpm_seal_thickness, edpm_seal_thickness, steel_z, 0.,
+                                                G4ThreeVector(0., -beam_thickness_2/2., 0.), 0);
+
+    G4double edpm_front_vol_   = 2*edpm_seal_thickness * steel_y * edpm_seal_thickness;
```
Same as above.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -343,123 +550,224 @@ namespace nexus {
-      	    vertex = front_roof_gen_->GenerateVertex("INSIDE");
-      	    vertex.setZ(vertex.z() - (shield_z_/2.+steel_thickness_+lead_thickness_/2.));
-      	    // std::cout<<"frontal -"<<std::endl;
+      	    vertex = ped_support_bottom_gen_->GenerateVertex("INSIDE");
```
Instead of generating the vertex twice, couldn't you just use the same vertex and shift it if the random number is less than 0.5? The same for the rest of the code and also in the `SHIELDING_STRUCT` generation. Please, tell me if I'm not seeing something here.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -343,123 +550,224 @@ namespace nexus {
+      if (rand<perc_edpm_front_vol_){ // front
+        if (G4UniformRand() < 0.5){
+          vertex = edpm_seal_front_gen_->GenerateVertex("INSIDE");
+          vertex.setZ(vertex.z() + (shield_z_/2. + steel_thickness_ + edpm_seal_thickness/2.));
```
The vertices are being generated into the `LEAD_BOX` volume, but this is not the intended behaviour, right? They should be generated just on the surface of the steel box, because they face the inner air.  I think the displacement should be 
`vertex.setZ(vertex.z() + (shield_z_/2. - edpm_seal_thickness/2.));`, do you agree?

gondiaz replied:
It is not still clear, since from the slides it seems that the steel is above the seal. Let's wait for Jordi's answer to finally set the exact position.

gondiaz replied:
After Jordi's confirmation, we move the EDPM vertex to be generated from inside the INNER_AIR.

paolafer commented on source/geometries/Next100Shielding.cc:
```diff
@@ -343,123 +550,224 @@ namespace nexus {
+      if (rand<perc_bubble_front_vol_){ // front
+        if (G4UniformRand() < 0.5){
+          vertex = bubble_seal_front_gen_->GenerateVertex("INSIDE");
+          vertex.setZ(vertex.z() + (support_beam_dist_/2. + support_front_dist_ + pedestal_front_beam_thickness/2. + bubble_seal_thickness/2.));
```
To simplify the code further and make it more readable, you can put the value to add or subtract in a different variable outside the `if (G4UniformRand() < 0.5)`. The same in `EDPM_SEAL`.

gondiaz committed:
Adjust trailing underscores to nexus convention

gondiaz committed:
Avoid duplicated code in vertex generators

gondiaz committed:
Try to keep <100 character line length

gondiaz committed:
Move EDPM seal vertex to INNER_AIR region

paolafer reviewed (approved):
This PR makes a thorough revision of the geometry of the external shielding, fixing some inaccuracies and bugs and checking all the dimensions with the drawings from Jordi. The code works as intended and has been reviewed in detail, so I approve it. Excellent job, @gondiaz!
